Expose an additional Table type when ResultType.DataTable is used

smoothdeveloper opened the pull request:
I think this is minimal firs step to get this feature done, I'm open for suggestions if you want me to address any concern.

I'd like to work on removing code duplication between `SqlProgrammabilityProvider` and `SqlCommandProvider` but I wonder if this is not already being done in V2 branch and I think this should be a separate PR.

Detail of actual changes to make the feature work:

* add DesignTime.GetDataTableType to encapsulate creation of Table type (might make the type name a parameter in later consolidation, see note)
* when ResultType.DataTable is used, add the ProvidedType (which was already of DataTable type) as a member to SqlCommandProvider provided type
* add a simple unit test which shows that we can specify the type explicitly (I've removed importing ProgrammabilityTest as it is only needed for a single SqlProgrammabilityProvider definition, better to not couple test modules too much)

Note: I've only changed SqlCommandProvider, but I noticed there is some code duplication already for the SqlClient provider itself in handling the DataRow and DataTable types.

#214

smoothdeveloper commented:
Can we get this reviewed/merged?

I'd also to take a look at exposing properties for each DataColumn on instance of the DataTable type, I think this will be very suitable.

dmitry-a-morozov commented:
- I checked out your PR and get compilation error:

Severity	Code	Description	Project	File	Line	Suppression State
Error		The type provider 'FSharp.Data.SqlProgrammabilityProvider' reported an error: type 'Table' was not added as a member to a declaring type	SqlClient.Tests	C:\Users\dmitry\Documents\GitHub\FSharp.Data.SqlClient\src\SqlClient.Tests\CreateCommand.fs	78	

- For consistency with static table types design ```Row``` should nested type of ```Table``` for example: ```GetArbitraryDataAsDataTable.Table.Row```. It will be breaking change.

smoothdeveloper commented:
@dmitry-a-morozov sorry about compile error, had some tests excluded from compilation as I don't have the adventure works db locally and it was taking long to compile.

The error is due to the fact I didn't address the static `CreateCommand` method, I'm looking into it now.

I'll also change Row to be a nested type as you mention for consistency.

Thanks for the review.

smoothdeveloper commented:
@dmitry-a-morozov I updated the PR, and ran the tests (now I have adventure works on my own machine, will be easier to contribute :smile:), please check again when time allows.

I've added a sample of referring to the `Table` and `Table.Row` types to the documentation where it felt relevant (where the `ResultType.DataTable` is used).

dmitry-a-morozov commented:
BTW, fake build script has task to deploy AdventureWorks database if missing
https://github.com/fsprojects/FSharp.Data.SqlClient/blob/master/build.fsx#L89

dmitry-a-morozov commented:
@smoothdeveloper Can you re-base against  master HEAD? thanks.

smoothdeveloper commented:
@dmitry-a-morozov updated, also fixed indentation to 4 spaces like the rest of codebase.

dmitry-a-morozov commented:
Commits merged as part of https://github.com/fsprojects/FSharp.Data.SqlClient/pull/218
